Some more documentation work

pitdicker opened the pull request:
- With `#![doc(test(attr(allow(unused_variables), deny(warnings))))]` we now make sure all doctest warnings do not get ignored, see https://github.com/rust-lang-nursery/rand/pull/363.
- I made small tweaks to the documentation: linkified some things, shortened the initial description of some structs, less or more documentation for re-exports.
- Better documentation and examples for `JitterRng`.
- Reordered the methods of `Rng`, so they show up in a somewhat more logical order in the documentation.
- Removed hard-coded links to `rand_core` on docs.rs. Because Rand re-exports the things that are linked to, we can just link to the documentation of the re-exports.

I think this is best to review per commit.

dhardy commented on src/distributions/range.rs:
```diff
@@ -88,6 +88,7 @@ impl<X: SampleRange> Distribution<X> for Range<X> {
 /// Helper trait for creating objects using the correct implementation of
 /// `RangeImpl` for the sampling type; this enables `Range::new(a, b)` to work.
 pub trait SampleRange: PartialOrd+Sized {
+    /// Actual `RangeImpl` implementation for `X`.
```
I'd prefer you didn't use 'actual'; try "The `RangeImpl` implementation supporting type `X`."

Also, it would make sense now to rename `X` → `T` and `T` to something else (perhaps `Impl`; it's not something users will see a lot). This is what Huon suggested I think.

pitdicker replied:
That sounds better, as always.

dhardy commented on src/lib.rs:
```diff
@@ -174,7 +182,9 @@
 
+#![deny(missing_docs)]
 #![deny(missing_debug_implementations)]
+#![doc(test(attr(allow(unused_variables), deny(warnings))))]
```
I generally prefer warnings to errors (`#![warn(...)]`) simply because new compiler versions may add warnings without considering backwards compatibility, thus breaking builds. In the latter case unfortunately we have no choice since warnings from examples are not shown; also "missing docs" is not something the compiler can break so that's fine.

pitdicker replied:
Good point, I'll keep that in mind!

dhardy commented on src/lib.rs:
```diff
@@ -209,15 +219,14 @@ pub use rand_core::{RngCore, BlockRngCore, CryptoRng, SeedableRng};
-#[cfg(feature="std")] pub use os::OsRng;
+#[doc(inline)] pub use jitter::JitterRng;
+#[cfg(feature="std")]
+#[doc(inline)] pub use os::OsRng;
```
Keep the OsRng import on a single line please; this causes chaos with diff tools.

dhardy commented on src/lib.rs:
```diff
@@ -489,61 +388,101 @@ pub trait Rng: RngCore {
     }
 
-    /// Return a bool with a 1 in n chance of true
+    /// Sample a new value, using the given distribution.
```
This commit is just reordering, and this is the diff tool getting confused?

pitdicker replied:
Yes, I made sure not to change anything but the order.

dhardy commented:
Needs rebase

pitdicker commented:
That rebase was harder than I thought :smile:. I removed the commit that reordered the Rng methods, and recreated it at the end, to make sure I did not lose changes during the rebase.

pitdicker commented:
Added three more commits, that may need some explanation.

`cargo-deadlinks` proved really useful to make sure all documentation links are valid. At least for `rand`, there are two links in `rand_core` I don't know what to do with. I have added it to the CI, but don't know if installing `cargo-deadlinks` will take too much time.

I set the dependency to `rand_core` to a path in the repro so it uses the fixed links. When `rand_core` 0.1 is released, everything should work out and `rand` master can again depend on the crates.io version. But it seems like a good idea to me to depend on a relative path on master sometimes.

The line "[`rand_core::impls`]: ../rand_core/impls/index.html" has the link go out of the `rand_core` crate, and then back in. This way it also works for all the re-exported functions in `rand`. At least locally and on GitHub pages. I added a section to `cargo.toml` to give hints to docs.rs ([documentation](https://github.com/onur/docs.rs/blob/master/src/docbuilder/metadata.rs)), so it hopefully builds `rand_core` together with the `rand_crate`.

Finally I tried to clean up the list of exported items in `libs.rs`. At least in the documentation by hiding things, not by breaking the API. A few more things are now directly available like `ReadRng` and `ReseedingRng`, instead of under a module with only one item.

Two exceptions: I broke `JitterRng` by only exporting the module (couldn't get the links to work from both places). We need to have the module public, because it also contains an error type. And I think not having it at the top level import can even be the right thing... Found one very recent dependency on GitHub that uses it directly.

Also the `reseeding` module is now not exported any more. We can basically consider it a rewrite, and all existing code using it will have to update. No point in making sure the export does not break :smile:.

One oddity: even when we build the documentation with "--all-features" `XorShiftRng` and `IsaacRng` don't show the implement `Serialize` and `Deserialize`. This is due to https://github.com/rust-lang/rust/issues/36922.

pitdicker commented:
> I'd have preferred it if you'd brought up the breaking changes in a separate issue/PR. 

Yes, that would have been better. It was kind of necessary to make the links checking work out, so it would have been hard to separate.

pitdicker commented on .travis.yml:
```diff
@@ -34,6 +34,8 @@ matrix:
         - cargo test --benches
         - cargo doc --no-deps --all --all-features
+        - cargo --list | egrep "^\s*deadlinks$" -q || cargo install cargo-deadlinks
+        - cargo deadlinks
```
Yes, only the documentation cross-links. But I suppose checking all links can be useful...

pitdicker commented:
Updated the commits.
